### PR TITLE
[Feature] random extension libraries

### PR DESCRIPTION
**Pull request by nobu**
Allows to inherit `Random` and implement other algorithms.

### CODE REVIEW

File: include/ruby/random.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
+#ifndef RUBY_RANDOM_H
+#define RUBY_RANDOM_H 1
+/**
+ * @file
+ * @author     $Author$
```
Comment (mrkn) [inline include/ruby/random.h:5]: We should stop using `$Author$` here.

File: ext/random/chacha/chacha.c [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+static
+#ifndef _MSC_VER
+const
+#endif
+rb_data_type_t random_chacha_type = {
+    "random/ChaCha",
+    {
+	rb_random_mark,
```
Comment (mrkn) [inline ext/random/chacha/chacha.c:53]: We use soft tabs instead of hard tabs now in newly added files.

File: include/ruby/random.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+typedef void rb_random_init_func(rb_random_t *, const uint32_t *, size_t);
+typedef unsigned int rb_random_get_int32_func(rb_random_t *);
+typedef void rb_random_get_bytes_func(rb_random_t *, void *, size_t);
+
+typedef struct {
+    size_t default_seed_bits;
+    rb_random_init_func *init;
+    rb_random_get_int32_func *get_int32;
+    rb_random_get_bytes_func *get_bytes;
+} rb_random_interface_t;
```
Comment (mrkn) [inline include/ruby/random.h:38]: Some random generators,  such as dSFMT, can generate real random numbers directly.
How to support directly generation of real random numbers in such generators?

File: include/ruby/random.h (hunk 1 above)
Comment (nobu) [reply include/ruby/random.h:5]: Remove that line?

File: include/ruby/random.h (hunk 3 above)
Comment (nobu) [reply include/ruby/random.h:38]: What signature does that function have?

File: include/ruby/random.h (hunk 1 above)
Comment (mrkn) [reply include/ruby/random.h:5]: I think we can drop this line.

File: include/ruby/random.h (hunk 3 above)
Comment (mrkn) [reply include/ruby/random.h:38]: The signature of the function that generates a real random number in `[0, 1)` is the same as `rb_random_get_int32_func`.

In the original implementation of dSFMT, the prototype is:

```c
double dsfmt_genrand_close_open(dsfmt_t *dsfmt);
```

You can find this line at https://github.com/MersenneTwister-Lab/dSFMT/blob/master/dSFMT.h#L206

Comment (mrkn) [reply include/ruby/random.h:38]: We can make `get_int32` is required, and `get_double` is optional.
If a given generator type provides `get_double` function, `random_real` can use `get_double`.

**Comment by shyouhei** [comment]
Don't roll your own crypto.

Huge :-1: to copy&pastel ChaCha in this repository.

**Comment by simi** [comment]
:mag: Is there Feature number available (aka redmine link) for this?

**Comment by mrkn** [comment]
We can remove `Random::ChaCha` from this pull-request.

**Comment by shyouhei** [comment]
@mrkn please do.  It's perfectly okay for you to have ChaCha as your private gem.  I'm just against such thing to be a part of this repository.

**Comment by mrkn** [comment]
@nobu Could you remove ChaCha from this pull-request?  I guess this pull-request mainly purpose to introduce the mechanism of implementing the original random number generation algorithms.  ChaCha shouldn't be important in this.

I'm interested in to write generators for the Monte Carlo method other than MersenneTwister, such as XorShift, dSFMT, and PCG.  This new mechanism looks good to me for such a purpose.

**Comment by nobu** [comment]
Of course, ChaCha was just an example, I needed to ensure it worked actually.
In fact, it hadn't woked on Windows.

File: include/ruby/random.h (hunk 3 above)
Comment (nobu) [reply include/ruby/random.h:37]: We need not only close-open range but also close-close range for inclusive `Range`.

**Comment by nobu** [comment]
Right now, keeping ChaCha to try to compile an extension on Windows.

**Comment by shyouhei** [comment]
Don't forget to delete ChaCha before pressing the merge button.

**Comment by nobu** [comment]
Added loop number generator instead of ChaCha, for the test.
Test for `get_real` would be needed too.

File: ext/-test-/random/loop.c [hunk 4]
```diff
@@ -0,0 +1,110 @@
+#include "ruby/random.h"
+
+static const uint32_t max_seeds = 1024;
+
+typedef struct {
+    rb_random_t base;
+    uint32_t num, idx, *buf;
+} rand_loop_t;
+
+static double loop_get_real(rb_random_t *rnd, int excl);
```
Comment (mrkn) [inline ext/-test-/random/loop.c:10]: @nobu How about adding `RB_RANDOM_INTERFACE_DECLARE_WITH_REAL` to declare functions with `get_real`?

File: ext/-test-/random/loop.c [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+RB_RANDOM_INTERFACE_DECLARE(loop)
+static const rb_random_interface_t random_loop_if = {
+    32,
+    RB_RANDOM_INTERFACE_DEFINE(loop)
+    loop_get_real,
```
Comment (mrkn) [inline ext/-test-/random/loop.c:16]: How about adding `RB_RANDOM_INTERFACE_DEFINE_WITH_REAL` same as above?

File: ext/-test-/random/loop.c [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    RUBY_TYPED_FREE_IMMEDIATELY
+};
+
+
+static VALUE
+loop_alloc(VALUE klass)
+{
+    rand_loop_t *rnd;
+    VALUE obj = TypedData_Make_Struct(klass, rand_loop_t, &random_loop_type, rnd);
+    rnd->base.seed = INT2FIX(0);
```
Comment (mrkn) [inline ext/-test-/random/loop.c:44]: @nobu Should we provide the initializer function of base structure?

**Pull request merged**